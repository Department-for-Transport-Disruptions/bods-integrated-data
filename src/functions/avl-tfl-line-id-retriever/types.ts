export type tflLineResponse = {
    $type?: string;
    id?: number;
    name?: string;
    modeName?: string;
    disruptions?: string;
    created?: string;
    modified?: string;
    lineStatuses?: string;
    routeSections?: string;
    serviceTypes?: string;
    crowding?: string;
};
