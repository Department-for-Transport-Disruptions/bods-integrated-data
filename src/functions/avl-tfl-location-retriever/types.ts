export type TflApiKeys = {
    general_api_key: string;
    live_vehicles_api_key: string;
};

export type RealTimeVehicleLocation = {
    producerRef?: string;
    vehicleRef?: string;
    vehicleName?: string;
    operatorRef?: string;
    monitored?: string;
    longitude?: number;
    latitude?: number;
    recordedAtTime?: string;
    bearing?: number;
    load?: number;
    passengerCount?: number;
    odometer?: number;
    headwayDeviation?: number;
    scheduleDeviation?: number;
    vehicleState?: number;
    nextStopPointId?: string;
    nextStopPointName?: string;
    previousStopPointId?: string;
    previousStopPointName?: string;
    lineRef?: string;
    publishedLineName?: string;
    directionRef?: number;
    originName?: string;
    originRef?: string;
    originAimedDepartureTime?: number;
    destinationName?: string;
    destinationRef?: string;
    vehicleJourneyRef?: string;
};

export type RealTimeVehicleLocationsApiResponse = {
    lines: {
        lineId: string;
        vehicles: RealTimeVehicleLocation[];
    }[];
};
