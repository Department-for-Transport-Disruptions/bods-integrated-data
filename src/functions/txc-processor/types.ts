import { NewCalendar, NewCalendarDate } from "@bods-integrated-data/shared/database";
import { JourneyPattern, VehicleJourney } from "@bods-integrated-data/shared/schema";

export type VehicleJourneyMapping = {
    vehicleJourney: VehicleJourney;
    routeId: number;
    serviceId: number;
    shapeId: string | null;
    tripId: string;
    serviceCode: string;
    journeyPattern?: JourneyPattern;
};

export type CalendarWithDates = {
    calendar: NewCalendar;
    calendarDates: Omit<NewCalendarDate, "service_id">[];
};
