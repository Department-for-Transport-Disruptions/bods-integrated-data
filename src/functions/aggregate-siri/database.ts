import { Database, Logger, notEmpty } from "@bods-integrated-data/shared";
import { avlSchema } from "@bods-integrated-data/shared/schema/siri.schema";
import { Kysely, sql } from "kysely";

export const getCurrentAvlData = async (db: Kysely<Database>, logger: Logger) => {
    logger.info("Getting data from avl table...");

    const avl = await db
        .selectFrom("avl")
        .distinctOn(["operatorRef", "vehicleRef"])
        .selectAll("avl")
        .orderBy(sql`"operatorRef", "vehicleRef", "responseTimeStamp" DESC`)
        .execute();

    const parsedAvl = avl.map((record) => {
        const parsedAvl = avlSchema.safeParse(record);

        if (!parsedAvl.success) {
            logger.warn(`Invalid avl data in database. ID: ${record.id}`);
            logger.warn(`${JSON.stringify(parsedAvl.error)}`);
            return null;
        }

        return parsedAvl.data;
    });

    return parsedAvl.filter(notEmpty);
};
