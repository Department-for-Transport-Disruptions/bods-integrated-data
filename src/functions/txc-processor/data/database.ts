import { logger } from "@baselime/lambda-logger";
import {
    Agency,
    Database,
    LocationType,
    NewCalendar,
    NewCalendarDate,
    NewRoute,
    NewStop,
} from "@bods-integrated-data/shared/database";
import { Operator, Service, TxcStop } from "@bods-integrated-data/shared/schema";
import { getRouteTypeFromServiceMode, notEmpty } from "@bods-integrated-data/shared/utils";
import { Kysely } from "kysely";
import { hasher } from "node-object-hash";

export const insertAgencies = async (dbClient: Kysely<Database>, operators: Operator[]) => {
    const agencyPromises = operators.map(async (operator) => {
        const existingAgency = await dbClient
            .selectFrom("agency")
            .selectAll()
            .where("noc", "=", operator.NationalOperatorCode)
            .executeTakeFirst();

        return dbClient
            .insertInto("agency_new")
            .values(
                existingAgency || {
                    name: operator.OperatorShortName,
                    noc: operator.NationalOperatorCode,
                    url: "",
                    registered_operator_ref: operator["@_id"],
                },
            )
            .onConflict((oc) => oc.column("noc").doUpdateSet({ name: operator.OperatorShortName }))
            .returningAll()
            .executeTakeFirst();
    });

    const agencyData = await Promise.all(agencyPromises);

    return agencyData.filter(notEmpty);
};

export const insertCalendar = async (
    dbClient: Kysely<Database>,
    calendarData: {
        calendar: NewCalendar;
        calendarDates: NewCalendarDate[];
    },
) => {
    const calendarHash = hasher().hash(calendarData);

    const insertedCalendar = await dbClient
        .insertInto("calendar_new")
        .values({ ...calendarData.calendar, calendar_hash: calendarHash })
        .onConflict((oc) => oc.column("calendar_hash").doUpdateSet({ ...calendarData.calendar }))
        .returningAll()
        .executeTakeFirst();

    if (!insertedCalendar?.id) {
        throw new Error("Calendar failed to insert");
    }

    await dbClient
        .insertInto("calendar_date_new")
        .values(
            calendarData.calendarDates.map((date) => ({
                date: date.date,
                exception_type: date.exception_type,
                service_id: insertedCalendar.id,
            })),
        )
        .onConflict((oc) => oc.doNothing())
        .execute();

    return insertedCalendar;
};

export const insertRoutes = async (dbClient: Kysely<Database>, service: Service, agencyData: Agency[]) => {
    const agency = agencyData.find((agency) => agency.registered_operator_ref === service.RegisteredOperatorRef);

    if (!agency) {
        logger.warn(`Unable to find agency with registered operator ref: ${service.RegisteredOperatorRef}`);
        return null;
    }

    const routeType = getRouteTypeFromServiceMode(service.Mode);

    const routePromises = service.Lines.Line.map(async (line) => {
        const existingRoute = await dbClient
            .selectFrom("route")
            .selectAll()
            .where("line_id", "=", line["@_id"])
            .executeTakeFirst();

        const newRoute: NewRoute = {
            agency_id: agency.id,
            route_short_name: line.LineName,
            route_long_name: "",
            route_type: routeType,
            line_id: line["@_id"],
        };

        return dbClient
            .insertInto("route_new")
            .values(existingRoute || newRoute)
            .onConflict((oc) =>
                oc.column("line_id").doUpdateSet({ route_short_name: line.LineName, route_type: routeType }),
            )
            .returningAll()
            .executeTakeFirst();
    });

    const routeData = await Promise.all(routePromises);

    return routeData.filter(notEmpty);
};

export const insertStops = async (dbClient: Kysely<Database>, stops: TxcStop[]) => {
    const platformCodes = ["BCS", "PLT", "FBT"];
    const atcoCodes = stops.map((stop) => stop.StopPointRef);

    const naptanStops = await dbClient
        .selectFrom("naptan_stop_new")
        .selectAll()
        .where("atco_code", "in", atcoCodes)
        .execute();

    const stopsToInsert = stops.map((stop): NewStop => {
        const naptanStop = naptanStops.find((s) => s.atco_code === stop.StopPointRef);

        return {
            id: stop.StopPointRef,
            wheelchair_boarding: 0,
            parent_station: "",

            ...(naptanStop
                ? {
                      stop_code: naptanStop.naptan_code,
                      stop_name: naptanStop.common_name || stop.CommonName,
                      stop_lat: naptanStop.latitude ? parseFloat(naptanStop.latitude) : stop.Location?.Latitude,
                      stop_lon: naptanStop.longitude ? parseFloat(naptanStop.longitude) : stop.Location?.Longitude,
                      location_type:
                          naptanStop.stop_type === "RSE" ? LocationType.RealStationEntrance : LocationType.None,
                      platform_code:
                          naptanStop.stop_type && platformCodes.includes(naptanStop.stop_type)
                              ? naptanStop.stop_type
                              : "",
                  }
                : {
                      stop_name: stop.CommonName,
                      stop_lat: stop.Location?.Latitude,
                      stop_lon: stop.Location?.Longitude,
                      location_type: LocationType.None,
                      platform_code: "",
                  }),
        };
    });

    await dbClient
        .insertInto("stop_new")
        .values(stopsToInsert)
        .onConflict((oc) => oc.column("id").doNothing())
        .returningAll()
        .executeTakeFirst();
};
