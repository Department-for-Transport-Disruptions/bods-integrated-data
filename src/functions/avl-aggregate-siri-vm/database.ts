import { logger } from "@baselime/lambda-logger";
import { KyselyDb } from "@bods-integrated-data/shared/database";

export const getCurrentAvlData = async (dbClient: KyselyDb) => {
    logger.info("Getting data from avl table...");

    return dbClient
        .selectFrom("avl")
        .distinctOn(["operator_ref", "vehicle_ref"])
        .selectAll("avl")
        .orderBy(["operator_ref", "vehicle_ref", "response_time_stamp desc"])
        .execute();
};
