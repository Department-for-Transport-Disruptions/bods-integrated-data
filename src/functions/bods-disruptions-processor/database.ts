import { KyselyDb } from "@bods-integrated-data/shared/database";

export const getAgencies = (dbClient: KyselyDb, nocs: string[]) => {
    return dbClient.selectFrom("agency").selectAll().where("noc", "in", nocs).execute();
};

export const getRoutes = (dbClient: KyselyDb, lineRefs: string[]) => {
    return dbClient
        .selectFrom("route")
        .selectAll()
        .where("route_short_name", "in", lineRefs)
        .orderBy("id desc")
        .execute();
};
