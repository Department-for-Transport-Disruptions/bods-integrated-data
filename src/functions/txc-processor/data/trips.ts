import { KyselyDb, NewTrip } from "@bods-integrated-data/shared/database";
import { notEmpty, getWheelchairAccessibilityFromVehicleType } from "@bods-integrated-data/shared/utils";
import { randomUUID } from "crypto";
import { insertTrips } from "./database";
import { VehicleJourneyMapping } from "../types";

export const processTrips = async (
    dbClient: KyselyDb,
    vehicleJourneyMappings: VehicleJourneyMapping[],
    filePath: string,
) => {
    const updatedVehicleJourneyMappings = structuredClone(vehicleJourneyMappings);

    const trips = vehicleJourneyMappings
        .map<NewTrip | null>((vehicleJourneyMapping, index) => {
            const { vehicleJourney, journeyPattern } = vehicleJourneyMapping;

            const tripId = randomUUID();
            updatedVehicleJourneyMappings[index].tripId = tripId;

            return {
                id: tripId,
                route_id: vehicleJourneyMapping.routeId,
                service_id: vehicleJourneyMapping.serviceId,
                block_id: vehicleJourney.Operational?.Block?.BlockNumber || "",
                shape_id: vehicleJourneyMapping.shapeId,
                trip_headsign: vehicleJourney.DestinationDisplay || journeyPattern?.DestinationDisplay || "",
                wheelchair_accessible: getWheelchairAccessibilityFromVehicleType(
                    vehicleJourney.Operational?.VehicleType,
                ),
                vehicle_journey_code: vehicleJourney.VehicleJourneyCode,
                ticket_machine_journey_code: vehicleJourney.Operational?.TicketMachine?.JourneyCode || "",
                file_path: filePath,
            };
        })
        .filter(notEmpty);

    if (trips.length > 0) {
        await insertTrips(dbClient, trips);
    }

    return updatedVehicleJourneyMappings;
};
