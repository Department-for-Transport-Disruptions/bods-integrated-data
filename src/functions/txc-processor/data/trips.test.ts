import { KyselyDb, NewTrip, Trip, WheelchairAccessibility } from "@bods-integrated-data/shared/database";
import { beforeEach, describe, expect, it, vi } from "vitest";
import { VehicleJourneyMapping } from "../types";
import * as databaseFunctions from "./database";
import { processTrips } from "./trips";

describe("trips", () => {
    let dbClient: KyselyDb;
    const insertTripsMock = vi.spyOn(databaseFunctions, "insertTrips");

    beforeEach(() => {
        vi.resetAllMocks();
    });

    it("inserts trips into the database", async () => {
        const vehicleJourneyMappings: VehicleJourneyMapping[] = [
            {
                routeId: 1,
                serviceId: 2,
                shapeId: "3",
                tripId: "",
                serviceCode: "test",
                vehicleJourney: {
                    LineRef: "5",
                    ServiceRef: "6",
                    JourneyPatternRef: "7",
                    VehicleJourneyCode: "8",
                    DepartureTime: "00:00:00",
                    Operational: {
                        Block: {
                            BlockNumber: "block1",
                        },
                        TicketMachine: {
                            JourneyCode: "journey1",
                        },
                    },
                    DestinationDisplay: "vjDisplay1",
                },
                journeyPattern: {
                    "@_id": "1",
                    DestinationDisplay: "jpDisplay1",
                    JourneyPatternSectionRefs: [],
                    Direction: "inbound",
                },
            },
            {
                routeId: 11,
                serviceId: 12,
                shapeId: "13",
                tripId: "",
                serviceCode: "test",
                vehicleJourney: {
                    LineRef: "15",
                    ServiceRef: "16",
                    JourneyPatternRef: "17",
                    VehicleJourneyCode: "18",
                    DepartureTime: "00:01:00",
                },
                journeyPattern: {
                    "@_id": "1",
                    JourneyPatternSectionRefs: [],
                },
            },
        ];

        const expectedTrips: NewTrip[] = [
            {
                id: expect.any(String) as string,
                route_id: 1,
                service_id: 2,
                block_id: "block1",
                shape_id: "3",
                trip_headsign: "vjDisplay1",
                wheelchair_accessible: WheelchairAccessibility.NoAccessibilityInformation,
                vehicle_journey_code: "8",
                ticket_machine_journey_code: "journey1",
                file_path: "",
                direction: "inbound",
            },
            {
                id: expect.any(String) as string,
                route_id: 11,
                service_id: 12,
                block_id: "",
                shape_id: "13",
                trip_headsign: "",
                wheelchair_accessible: WheelchairAccessibility.NoAccessibilityInformation,
                vehicle_journey_code: "18",
                ticket_machine_journey_code: "",
                file_path: "",
                direction: "",
            },
        ];

        insertTripsMock.mockImplementation((_dbClient, trips) => Promise.resolve(trips) as Promise<Trip[]>);

        const updatedVehicleJourneyMappings = await processTrips(dbClient, vehicleJourneyMappings, "");

        expect(insertTripsMock).toHaveBeenCalledWith(dbClient, expectedTrips);
        expect(updatedVehicleJourneyMappings[0].tripId).toEqual(expectedTrips[0].id);
        expect(updatedVehicleJourneyMappings[1].tripId).toEqual(expectedTrips[1].id);
    });

    it("uses the journey pattern destination display when the vehicle journey destination display is omitted", async () => {
        const vehicleJourneyMappings: VehicleJourneyMapping[] = [
            {
                routeId: 1,
                serviceId: 2,
                shapeId: "3",
                tripId: "",
                serviceCode: "test",
                vehicleJourney: {
                    LineRef: "5",
                    ServiceRef: "6",
                    JourneyPatternRef: "7",
                    VehicleJourneyCode: "8",
                    DepartureTime: "00:00:00",
                    Operational: {
                        Block: {
                            BlockNumber: "block1",
                        },
                        TicketMachine: {
                            JourneyCode: "journey1",
                        },
                    },
                },
                journeyPattern: {
                    "@_id": "1",
                    DestinationDisplay: "jpDisplay1",
                    JourneyPatternSectionRefs: [],
                },
            },
        ];

        const expectedTrips: NewTrip[] = [
            {
                id: expect.any(String) as string,
                route_id: 1,
                service_id: 2,
                block_id: "block1",
                shape_id: "3",
                trip_headsign: "jpDisplay1",
                wheelchair_accessible: WheelchairAccessibility.NoAccessibilityInformation,
                vehicle_journey_code: "8",
                ticket_machine_journey_code: "journey1",
                file_path: "",
                direction: "",
            },
        ];

        insertTripsMock.mockImplementation((_dbClient, trips) => Promise.resolve(trips) as Promise<Trip[]>);

        await processTrips(dbClient, vehicleJourneyMappings, "");

        expect(insertTripsMock).toHaveBeenCalledWith(dbClient, expectedTrips);
    });

    it("doesn't insert trips into the database when the vehicle journey mapping is empty", async () => {
        await processTrips(dbClient, [], "");
        expect(insertTripsMock).not.toHaveBeenCalled();
    });
});
