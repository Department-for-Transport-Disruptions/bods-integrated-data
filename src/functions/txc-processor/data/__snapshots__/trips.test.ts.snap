// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`trips > inserts trips into the database 1`] = `
[
  {
    "blockId": "",
    "journeyPattern": {
      "@_id": "1",
      "JourneyPatternSectionRefs": [],
    },
    "routeId": 11,
    "serviceCode": "test",
    "serviceId": 12,
    "shapeId": "13",
    "tripId": "VJ668c580888094ff9bad05521b7b82f58e8f31ed0",
    "vehicleJourney": {
      "@_RevisionNumber": "2",
      "DepartureDayShift": 1,
      "DepartureTime": "00:01:00",
      "JourneyPatternRef": "17",
      "LineRef": "15",
      "ServiceRef": "16",
      "VehicleJourneyCode": "18",
    },
  },
  {
    "blockId": "ad4b5661e10bc4d2f9209e1e1f0083bc4112d31b",
    "journeyPattern": {
      "@_id": "1",
      "DestinationDisplay": "jpDisplay1",
      "Direction": "inbound",
      "JourneyPatternSectionRefs": [],
    },
    "routeId": 1,
    "serviceCode": "test",
    "serviceId": 2,
    "shapeId": "3",
    "tripId": "VJ28c521ca45f8b29027531375bd76dec15ec90dd5",
    "vehicleJourney": {
      "DepartureTime": "23:59:00",
      "DestinationDisplay": "vjDisplay1",
      "JourneyPatternRef": "7",
      "LineRef": "5",
      "OperatingProfile": {
        "RegularDayType": {
          "DaysOfWeek": {
            "Monday": "",
          },
        },
      },
      "Operational": {
        "Block": {
          "BlockNumber": "block1",
        },
        "TicketMachine": {
          "JourneyCode": "journey1",
        },
      },
      "ServiceRef": "6",
      "VehicleJourneyCode": "8",
    },
  },
]
`;
